pkg/controller/plan/adapter/ocp/builder.go

```diff
@@ -266,33 +268,47 @@ func (r *Builder) mapDisks(sourceVm *cnv.VirtualMachine, targetVmSpec *cnv.Virtu
 		}
 	}
 
+	diskMap := createDiskMap(sourceVm, pvcMap, vmRef)
+	configMaps, secrets := r.createEnvMaps(sourceVm, vmRef)
+
+	// Clear original disks and volumes, will be required for other mapped devices later
+	targetVmSpec.Template.Spec.Domain.Devices.Disks = []cnv.Disk{}
+	targetVmSpec.Template.Spec.Volumes = []cnv.Volume{}
+
+	r.mapPVCsToTarget(targetVmSpec, persistentVolumeClaims, diskMap)
+	r.mapConfigMapsToTarget(targetVmSpec, configMaps, diskMap)
+	r.mapSecretsToTarget(targetVmSpec, secrets, diskMap)
+}
+
+func createDiskMap(sourceVm *cnv.VirtualMachine, pvcMap map[string]*core.PersistentVolumeClaim, vmRef ref.Ref) map[string]*cnv.Disk {
 	diskMap := make(map[string]*cnv.Disk)
-	for i := range sourceVm.Spec.Template.Spec.Domain.Devices.Disks {
-		disk := &sourceVm.Spec.Template.Spec.Domain.Devices.Disks[i]
-		// Find matching volume
-		// TODO: make this better
+
+	for _, disk := range sourceVm.Spec.Template.Spec.Domain.Devices.Disks {
+		currentDisk := disk
```

Comment:
isn't `currentDisk` redundant?

Reply:
no because of the reference, otherwise we'll only get the last value, the other option is to use the index on the array

Reply:
doesn't it (the reference) change on each iteration?

Reply:
the reference of disk or currentDisk?
currentDisk is a new variable on each iteration so we append a different memory location every time

Reply:
I meant the `disk` variable but I see now - the values from the range are placed at the same address that `disk` is initially allocated on so you need to duplicate it, ack